src/const_eval.md

```diff
@@ -81,11 +81,127 @@ r[const-eval.const-expr.builtin-arith-logic]
   operators used on integer and floating point types, `bool`, and `char`.
 
 r[const-eval.const-expr.borrows]
-* All forms of [borrow]s, including raw borrows, with one limitation:
-  mutable borrows and shared borrows to values with interior mutability
-  are only allowed to refer to *transient* places or to *static* places. A place is *transient*
-  if its lifetime is strictly contained inside the current [const context].
-  A place is *static* if it is a `static` item or a [promoted expression].
+* All forms of [borrow]s, including raw borrows, except:
+
+  * Mutable borrows of expressions whose temporary scopes would be extended (see [temporary lifetime extension]) to the end of the program.
+  * Shared borrows of expressions whose temporary scopes would be extended to the end of the program when those expressions result in values with interior mutability.
```

Comment:
Personally I'd prefer to avoid the redundancy here and spell this out a bit differently. For instance, we could say something like

* All forms of [borrow]s, including raw borrows, except for those that satisfy all of the following:
  * The borrows is either a mutable borrows, or it is a shared borrow and the borrowed expression results in a value with interior mutability.
  * The borrowed expression is a temporary that has its scope extended to the end of the program (see [destructors.scope.lifetime-extension.static]).

But this is just editorial so :shrug:

Reply:
Came up with something here that achieves this with good readability.